fatoptimizer/inline.py

```diff
@@ -1,35 +1,128 @@
 import ast
 
-from .tools import OptimizerStep, pretty_dump
-#from .specialized import BuiltinGuard
+from .tools import OptimizerStep, pretty_dump, NodeTransformer, NodeVisitor
 
+class Checker(NodeVisitor):
+    '''Gather a list of problems that would prevent inlining a function.'''
+    def __init__(self):
+        self.problems = []
+
+    def visit_Call(self, node):
+        # Reject explicit attempts to use locals()
+        # TODO: detect uses via other names
+        if isinstance(node.func, ast.Name):
+            if node.func.id == 'locals':
+                self.problems.append('use of locals()')
+
+def locate_kwarg(funcdef, name):
+    '''Get the index of an argument of funcdef by name.'''
+    for idx, arg in enumerate(funcdef.args.args):
+        if arg.arg == name:
+            return idx
+    raise ValueError('argument %r not found' % name)
+
+class RenameVisitor(NodeTransformer):
+    def __init__(self, callsite, inlinable, actual_pos_args):
+        assert callsite.starargs is None
+        assert callsite.kwargs is None
+        assert inlinable.args.vararg is None
+        assert inlinable.args.kwonlyargs == []
+        assert inlinable.args.kw_defaults == []
+        assert inlinable.args.kwarg is None
+        assert inlinable.args.defaults == []
+
+        # Mapping from name in callee to node in caller
+        self.remapping = {}
+        for formal, actual in zip(inlinable.args.args, actual_pos_args):
+            self.remapping[formal.arg] = actual
+
+    def visit_Name(self, node):
+        if node.id in self.remapping:
+            return self.remapping[node.id]
+        return node
+
```

Comment:
coding style: I'm now trying to use two empty lines between functions, classes, etc. (PEP 8)